File: src/domain/models/ticket_plan.py
"""TicketPlan aggregate root for the Ticket Pipeline bounded context.

Generates dependency-ordered beads tickets from a DomainModel's bounded
contexts and aggregate designs. Detail level is driven by subdomain
classification (complexity budget).

Invariants:
1. One epic per bounded context -- no duplicates.
2. Dependency order is topologically sorted (foundation first).
3. Circular dependencies are detected and rejected.
4. Core subdomains produce FULL-detail tickets.
5. No cross-BC dependency references (tickets only depend within their BC
   or on tickets in upstream BCs per context relationships).
"""

from __future__ import annotations

import uuid
from collections import deque
from typing import TYPE_CHECKING

from src.domain.models.domain_values import AggregateDesign
from src.domain.models.errors import InvariantViolationError
from src.domain.models.ticket_values import (
    DependencyOrder,
    GeneratedEpic,
    GeneratedTicket,
    TicketDetailLevel,
    classify_tier,
    tier_to_detail_level,
)
from src.domain.services.ticket_detail_renderer import TicketDetailRenderer

if TYPE_CHECKING:
    from src.domain.events.ticket_events import TicketPlanApproved
    from src.domain.models.domain_model import DomainModel
    from src.domain.models.domain_values import SubdomainClassification
    from src.domain.models.stack_profile import StackProfile


class TicketPlan:
    """Aggregate root: generates and manages a ticket plan from a DomainModel.

    Attributes:
        plan_id: Unique identifier for this plan.
    """

    def __init__(self) -> None:
        self.plan_id: str = str(uuid.uuid4())
        self._epics: list[GeneratedEpic] = []
        self._tickets: list[GeneratedTicket] = []
        self._dependency_order: DependencyOrder | None = None
        self._events: list[TicketPlanApproved] = []
        self._approved: bool = False
        self._profile: StackProfile | None = None

    # -- Properties -----------------------------------------------------------

    @property
    def epics(self) -> tuple[GeneratedEpic, ...]:
        """All generated epics (defensive copy)."""
        return tuple(self._epics)

    @property
    def tickets(self) -> tuple[GeneratedTicket, ...]:
        """All generated tickets (defensive copy)."""
        return tuple(self._tickets)

    @property
    def dependency_order(self) -> DependencyOrder | None:
        """Topologically sorted ticket execution order, or None if not computed."""
        return self._dependency_order

    @property
    def events(self) -> tuple[TicketPlanApproved, ...]:
        """Domain events produced by this aggregate (defensive copy)."""
        return tuple(self._events)

    # -- Commands -------------------------------------------------------------

    def generate_plan(self, model: DomainModel, profile: StackProfile | None = None) -> None:
        """Generate epics and tickets from a finalized DomainModel.

        Creates one epic per bounded context, then generates tickets per
        aggregate with detail level driven by subdomain classification.
        Generic BCs with no aggregates get a single stub ticket.

        Args:
            model: A finalized DomainModel with classified bounded contexts.
            profile: Stack profile for quality gate rendering. Defaults to
                PythonUvProfile for backward compatibility.

        Raises:
            InvariantViolationError: If model has no bounded contexts,
                any BC lacks classification, or a cycle is detected.
        """
        if self._approved:
            msg = "Cannot regenerate plan on an approved TicketPlan"
            raise InvariantViolationError(msg)

        if profile is None:
            from src.domain.models.stack_profile import PythonUvProfile

            profile = PythonUvProfile()
        self._profile = profile

        contexts = model.bounded_contexts
        if not contexts:
            msg = "No bounded contexts to generate tickets for"
            raise InvariantViolationError(msg)

        self._epics.clear()
        self._tickets.clear()

        # Build a lookup of aggregates by context name
        aggregates_by_context: dict[str, list[AggregateDesign]] = {}
        for agg in model.aggregate_designs:
            aggregates_by_context.setdefault(agg.context_name, []).append(agg)

        # Build upstream lookup from context relationships
        upstream_contexts: dict[str, set[str]] = {}
        for rel in model.context_relationships:
            upstream_contexts.setdefault(rel.downstream, set()).add(rel.upstream)

        # Track epic IDs by context name for cross-referencing
        epic_id_by_context: dict[str, str] = {}

        for bc in contexts:
            if bc.classification is None:
                msg = f"Bounded context '{bc.name}' has no subdomain classification"
                raise InvariantViolationError(msg)

            epic_id = str(uuid.uuid4())
            epic_id_by_context[bc.name] = epic_id

            self._epics.append(
                GeneratedEpic(
                    epic_id=epic_id,
                    title=f"{bc.name} Epic",
                    description=(
                        f"Implement the {bc.name} bounded context "
                        f"({bc.classification.value} subdomain)."
                    ),
                    bounded_context_name=bc.name,
                    classification=bc.classification,
                )
            )

            detail_level = TicketDetailLevel.from_classification(bc.classification)
            context_aggregates = aggregates_by_context.get(bc.name, [])

            if not context_aggregates:
                # Generic/Supporting BCs with no aggregates get a stub ticket
                stub_agg = AggregateDesign(
                    name=bc.name,
                    context_name=bc.name,
                    root_entity=bc.name,
                )
                stub_description = TicketDetailRenderer.render(
                    stub_agg, TicketDetailLevel.STUB, profile
                )
                self._tickets.append(
                    GeneratedTicket(
                        ticket_id=str(uuid.uuid4()),
                        title=f"Integrate {bc.name} boundary",
                        description=stub_description,
                        detail_level=TicketDetailLevel.STUB,
                        epic_id=epic_id,
                        bounded_context_name=bc.name,
                        aggregate_name=bc.name,
                    )
                )
            else:
                for agg in context_aggregates:
                    description = TicketDetailRenderer.render(agg, detail_level, profile)
                    self._tickets.append(
                        GeneratedTicket(
                            ticket_id=str(uuid.uuid4()),
                            title=f"Implement {agg.name} aggregate",
                            description=description,
                            detail_level=detail_level,
                            epic_id=epic_id,
                            bounded_context_name=bc.name,
                            aggregate_name=agg.name,
                        )
                    )

        # Assign cross-BC dependencies based on context relationships
        self._assign_cross_bc_dependencies(upstream_contexts, epic_id_by_context)

        # Compute topological order (INV2, INV3)
        self._dependency_order = self._compute_dependency_order()

        # Two-tier generation: compute depths and reclassify (alty-2j7.11)
        # classification is guaranteed non-None (validated at line 127)
        classification_by_context: dict[str, SubdomainClassification] = {
            bc.name: bc.classification  # type: ignore[misc]
            for bc in contexts
        }
        self._reclassify_by_depth(classification_by_context, profile)

    def preview(self) -> str:
        """Return a human-readable preview of the generated plan.

        Returns:
            Summary showing epic count and ticket counts by detail level.

        Raises:
            InvariantViolationError: If no plan has been generated.
        """
        if not self._epics:
            msg = "No plan generated yet -- call generate_plan() first"
            raise InvariantViolationError(msg)

        full_count = sum(1 for t in self._tickets if t.detail_level == TicketDetailLevel.FULL)
        standard_count = sum(
            1 for t in self._tickets if t.detail_level == TicketDetailLevel.STANDARD
        )
        stub_count = sum(1 for t in self._tickets if t.detail_level == TicketDetailLevel.STUB)

        lines: list[str] = [
            f"Ticket Plan: {self.plan_id}",
            f"Epics: {len(self._epics)}",
            f"Tickets: {len(self._tickets)} "
            f"(FULL={full_count}, STANDARD={standard_count}, STUB={stub_count})",
            "",
        ]

        for epic in self._epics:
            epic_tickets = [t for t in self._tickets if t.epic_id == epic.epic_id]
            lines.append(f"  {epic.title} ({epic.classification.value.upper()}):")
            lines.extend(
                f"    - [{ticket.detail_level.value}] {ticket.title}" for ticket in epic_tickets
            )
            lines.append("")

        return "\n".join(lines)

    def promote_stub(self, ticket_id: str, profile: StackProfile | None = None) -> None:
        """Promote a STUB ticket to FULL detail.

        Args:
            ticket_id: ID of the stub ticket to promote.
            profile: Stack profile for quality gate rendering. Uses the
                profile from generate_plan() if not provided.

        Raises:
            InvariantViolationError: If ticket not found or not a STUB.
        """
        for i, ticket in enumerate(self._tickets):
            if ticket.ticket_id == ticket_id:
                if ticket.detail_level != TicketDetailLevel.STUB:
                    msg = (
                        f"Ticket '{ticket_id}' is {ticket.detail_level.value}, "
                        f"not STUB -- cannot promote"
                    )
                    raise InvariantViolationError(msg)

                # Find or create aggregate design for re-rendering
                resolved_profile = profile or self._profile
                if resolved_profile is None:
                    from src.domain.models.stack_profile import PythonUvProfile

                    resolved_profile = PythonUvProfile()
                agg = AggregateDesign(
                    name=ticket.aggregate_name,
                    context_name=ticket.bounded_context_name,
                    root_entity=ticket.aggregate_name,
                )
                new_description = TicketDetailRenderer.render(
                    agg, TicketDetailLevel.FULL, resolved_profile
                )
                self._tickets[i] = GeneratedTicket(
                    ticket_id=ticket.ticket_id,
                    title=ticket.title,
                    description=new_description,
                    detail_level=TicketDetailLevel.FULL,
                    epic_id=ticket.epic_id,
                    bounded_context_name=ticket.bounded_context_name,
                    aggregate_name=ticket.aggregate_name,
                    dependencies=ticket.dependencies,
                    depth=ticket.depth,
                )
                return

        msg = f"Ticket '{ticket_id}' not found"
        raise InvariantViolationError(msg)

    def approve(
        self,
        approved_ids: tuple[str, ...] | None = None,
    ) -> None:
        """Approve the plan (all or a subset), emitting TicketPlanApproved.

        Args:
            approved_ids: If provided, only these ticket IDs are approved.
                          If None, all tickets are approved.

        Raises:
            InvariantViolationError: If plan already approved, has no tickets,
                or approved_ids contains unknown ticket IDs.
        """
        if self._approved:
            msg = "Plan already approved"
            raise InvariantViolationError(msg)

        if not self._tickets:
            msg = "Cannot approve plan with no tickets"
            raise InvariantViolationError(msg)

        all_ids = {t.ticket_id for t in self._tickets}

        if approved_ids is None:
            final_approved = tuple(t.ticket_id for t in self._tickets)
            final_dismissed: tuple[str, ...] = ()
        else:
            unknown = set(approved_ids) - all_ids
            if unknown:
                msg = f"Unknown ticket IDs: {', '.join(sorted(unknown))}"
                raise InvariantViolationError(msg)
            final_approved = approved_ids
            final_dismissed = tuple(tid for tid in all_ids if tid not in set(approved_ids))

        self._approved = True

        from src.domain.events.ticket_events import TicketPlanApproved

        self._events.append(
            TicketPlanApproved(
                plan_id=self.plan_id,
                approved_ticket_ids=final_approved,
                dismissed_ticket_ids=final_dismissed,
            )
        )

    def promotion_eligible_ids(self, resolved_ids: frozenset[str]) -> frozenset[str]:
        """Return IDs of STUB tickets whose dependencies are all resolved.

        Args:
            resolved_ids: Ticket IDs that have been completed/resolved.

        Returns:
            Frozenset of ticket IDs eligible for promotion.
        """
        eligible: set[str] = set()
        for ticket in self._tickets:
            if ticket.detail_level != TicketDetailLevel.STUB:
                continue
            if not ticket.dependencies:
                continue
            if all(dep_id in resolved_ids for dep_id in ticket.dependencies):
                eligible.add(ticket.ticket_id)
        return frozenset(eligible)

    # -- Private helpers ------------------------------------------------------

    def _compute_ticket_depths(self) -> dict[str, int]:
        """Compute depth for each ticket using topological order.

        Depth 0 = tickets with no in-plan dependencies (roots).
        Depth N = 1 + max(dependency depths).

        Requires ``_dependency_order`` to be computed first.
        """
        if self._dependency_order is None:
            msg = "Dependency order must be computed before depths"
            raise InvariantViolationError(msg)

        all_ids = {t.ticket_id for t in self._tickets}
        deps_map = {
            t.ticket_id: [d for d in t.dependencies if d in all_ids]
            for t in self._tickets
        }
        depths: dict[str, int] = {}

        for tid in self._dependency_order.ordered_ids:
            dep_depths = [depths[d] for d in deps_map[tid] if d in depths]
            depths[tid] = (max(dep_depths) + 1) if dep_depths else 0

        return depths

    def _reclassify_by_depth(
        self,
        classification_by_context: dict[str, SubdomainClassification],
        profile: StackProfile,
    ) -> None:
        """Reclassify ticket detail levels using depth-based two-tier rules.

        Near-term tickets (depth ≤2) keep their classification-based detail.
        Far-term tickets (depth >2) are downgraded to STUB.
        Core tickets are always FULL regardless of depth.
        """
        depths = self._compute_ticket_depths()
        updated: list[GeneratedTicket] = []

        for ticket in self._tickets:
            depth = depths.get(ticket.ticket_id, 0)
            classification = classification_by_context[ticket.bounded_context_name]
            tier = classify_tier(depth, classification)
            new_level = tier_to_detail_level(tier, classification)

            if new_level != ticket.detail_level:
                # Re-render description at new detail level
                agg = AggregateDesign(
                    name=ticket.aggregate_name,
                    context_name=ticket.bounded_context_name,
                    root_entity=ticket.aggregate_name,
                )
                new_description = TicketDetailRenderer.render(agg, new_level, profile)
                updated.append(
                    GeneratedTicket(
                        ticket_id=ticket.ticket_id,
                        title=ticket.title,
                        description=new_description,
                        detail_level=new_level,
                        epic_id=ticket.epic_id,
                        bounded_context_name=ticket.bounded_context_name,
                        aggregate_name=ticket.aggregate_name,
                        dependencies=ticket.dependencies,
                        depth=depth,
                    )
                )
            else:
                # Keep same ticket but add depth
                updated.append(
                    GeneratedTicket(
                        ticket_id=ticket.ticket_id,
                        title=ticket.title,
                        description=ticket.description,
                        detail_level=ticket.detail_level,
                        epic_id=ticket.epic_id,
                        bounded_context_name=ticket.bounded_context_name,
                        aggregate_name=ticket.aggregate_name,
                        dependencies=ticket.dependencies,
                        depth=depth,
                    )
                )

        self._tickets = updated

    def _assign_cross_bc_dependencies(
        self,
        upstream_contexts: dict[str, set[str]],
        epic_id_by_context: dict[str, str],
    ) -> None:
        """Assign dependencies to tickets based on context relationships.

        Tickets in downstream BCs depend on tickets in upstream BCs.
        """
        # Build a map of context name -> list of ticket IDs
        tickets_by_context: dict[str, list[str]] = {}
        for ticket in self._tickets:
            tickets_by_context.setdefault(ticket.bounded_context_name, []).append(ticket.ticket_id)

        # For each downstream context, make its tickets depend on upstream tickets
        updated_tickets: list[GeneratedTicket] = []
        for ticket in self._tickets:
            ctx = ticket.bounded_context_name
            upstream_names = upstream_contexts.get(ctx, set())
            dep_ids: list[str] = list(ticket.dependencies)
            for upstream_name in upstream_names:
                upstream_ticket_ids = tickets_by_context.get(upstream_name, [])
                dep_ids.extend(upstream_ticket_ids)
            if dep_ids != list(ticket.dependencies):
                updated_tickets.append(
                    GeneratedTicket(
                        ticket_id=ticket.ticket_id,
                        title=ticket.title,
                        description=ticket.description,
                        detail_level=ticket.detail_level,
                        epic_id=ticket.epic_id,
                        bounded_context_name=ticket.bounded_context_name,
                        aggregate_name=ticket.aggregate_name,
                        dependencies=tuple(dep_ids),
                    )
                )
            else:
                updated_tickets.append(ticket)
        self._tickets = updated_tickets

    def _compute_dependency_order(self) -> DependencyOrder:
        """Compute topological sort of tickets using Kahn's algorithm.

        Returns:
            DependencyOrder with tickets in dependency-safe execution order.

        Raises:
            InvariantViolationError: If a dependency cycle is detected.
        """
        # Build adjacency and in-degree
        all_ids = {t.ticket_id for t in self._tickets}
        in_degree: dict[str, int] = {tid: 0 for tid in all_ids}
        dependents: dict[str, list[str]] = {tid: [] for tid in all_ids}

        for ticket in self._tickets:
            for dep_id in ticket.dependencies:
                if dep_id in all_ids:
                    in_degree[ticket.ticket_id] += 1
                    dependents[dep_id].append(ticket.ticket_id)

        # Kahn's algorithm
        queue: deque[str] = deque()
        for tid, degree in in_degree.items():
            if degree == 0:
                queue.append(tid)

        ordered: list[str] = []
        while queue:
            current = queue.popleft()
            ordered.append(current)
            for dependent in dependents[current]:
                in_degree[dependent] -= 1
                if in_degree[dependent] == 0:
                    queue.append(dependent)

        if len(ordered) != len(all_ids):
            msg = "Circular dependency detected in ticket plan"
            raise InvariantViolationError(msg)

        return DependencyOrder(ordered_ids=tuple(ordered))

File: internal/ticket/domain/ticket_plan.go
package domain

import (
	"fmt"
	"strings"

	"github.com/alty-cli/alty/internal/shared/domain/ddd"
	"github.com/alty-cli/alty/internal/shared/domain/identity"
	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
)

// TicketPlan is the aggregate root for the ticket pipeline.
type TicketPlan struct {
	profile         vo.StackProfile
	dependencyOrder *DependencyOrder
	planID          string
	epics           []GeneratedEpic
	tickets         []GeneratedTicket
	events          []TicketPlanApproved
	approved        bool
}

// NewTicketPlan creates a new TicketPlan aggregate root.
func NewTicketPlan() *TicketPlan {
	return &TicketPlan{planID: identity.NewID()}
}

// PlanID returns the plan identifier.
func (p *TicketPlan) PlanID() string { return p.planID }

// Epics returns a defensive copy of all generated epics.
func (p *TicketPlan) Epics() []GeneratedEpic {
	out := make([]GeneratedEpic, len(p.epics))
	copy(out, p.epics)
	return out
}

// Tickets returns a defensive copy of all generated tickets.
func (p *TicketPlan) Tickets() []GeneratedTicket {
	out := make([]GeneratedTicket, len(p.tickets))
	copy(out, p.tickets)
	return out
}

// DependencyOrder returns the topological order, or nil if not computed.
func (p *TicketPlan) DependencyOrder() *DependencyOrder {
	if p.dependencyOrder == nil {
		return nil
	}
	d := *p.dependencyOrder
	return &d
}

// Events returns a defensive copy of domain events.
func (p *TicketPlan) Events() []TicketPlanApproved {
	out := make([]TicketPlanApproved, len(p.events))
	copy(out, p.events)
	return out
}

// GeneratePlan generates epics and tickets from a finalized DomainModel.
func (p *TicketPlan) GeneratePlan(model *ddd.DomainModel, profile vo.StackProfile) error {
	if p.approved {
		return fmt.Errorf("cannot regenerate plan on an approved TicketPlan: %w",
			domainerrors.ErrInvariantViolation)
	}
	if profile == nil {
		profile = vo.PythonUvProfile{}
	}
	p.profile = profile

	contexts := model.BoundedContexts()
	if len(contexts) == 0 {
		return fmt.Errorf("no bounded contexts to generate tickets for: %w",
			domainerrors.ErrInvariantViolation)
	}

	p.epics = nil
	p.tickets = nil

	// Build lookup of aggregates by context name.
	aggsByCtx := map[string][]vo.AggregateDesign{}
	for _, agg := range model.AggregateDesigns() {
		aggsByCtx[agg.ContextName()] = append(aggsByCtx[agg.ContextName()], agg)
	}

	// Build upstream lookup from context relationships.
	upstreamCtxs := map[string]map[string]bool{}
	for _, rel := range model.ContextRelationships() {
		if upstreamCtxs[rel.Downstream()] == nil {
			upstreamCtxs[rel.Downstream()] = map[string]bool{}
		}
		upstreamCtxs[rel.Downstream()][rel.Upstream()] = true
	}

	epicIDByCtx := map[string]string{}

	for _, bc := range contexts {
		if bc.Classification() == nil {
			return fmt.Errorf("bounded context '%s' has no subdomain classification: %w",
				bc.Name(), domainerrors.ErrInvariantViolation)
		}

		epicID := identity.NewID()
		epicIDByCtx[bc.Name()] = epicID
		classification := *bc.Classification()

		p.epics = append(p.epics, NewGeneratedEpic(
			epicID,
			bc.Name()+" Epic",
			fmt.Sprintf("Implement the %s bounded context (%s subdomain).",
				bc.Name(), string(classification)),
			bc.Name(),
			classification,
		))

		detailLevel := vo.DetailLevelFromClassification(classification)
		ctxAggs := aggsByCtx[bc.Name()]

		if len(ctxAggs) == 0 {
			stubAgg := vo.NewAggregateDesign(bc.Name(), bc.Name(), bc.Name(), nil, nil, nil, nil)
			stubDesc := RenderTicketDetail(stubAgg, vo.TicketDetailStub, profile)
			p.tickets = append(p.tickets, NewGeneratedTicket(
				identity.NewID(),
				"Integrate "+bc.Name()+" boundary",
				stubDesc,
				vo.TicketDetailStub,
				epicID, bc.Name(), bc.Name(), nil, 0,
			))
		} else {
			for _, agg := range ctxAggs {
				desc := RenderTicketDetail(agg, detailLevel, profile)
				p.tickets = append(p.tickets, NewGeneratedTicket(
					identity.NewID(),
					"Implement "+agg.Name()+" aggregate",
					desc,
					detailLevel,
					epicID, bc.Name(), agg.Name(), nil, 0,
				))
			}
		}
	}

	p.assignCrossBCDependencies(upstreamCtxs)
	order, err := p.computeDependencyOrder()
	if err != nil {
		return err
	}
	p.dependencyOrder = &order

	classificationByCtx := map[string]vo.SubdomainClassification{}
	for _, bc := range contexts {
		classificationByCtx[bc.Name()] = *bc.Classification()
	}
	p.reclassifyByDepth(classificationByCtx, profile)
	return nil
}

// Preview returns a human-readable preview of the generated plan.
func (p *TicketPlan) Preview() (string, error) {
	if len(p.epics) == 0 {
		return "", fmt.Errorf("no plan generated yet -- call GeneratePlan() first: %w",
			domainerrors.ErrInvariantViolation)
	}

	fullCount, stdCount, stubCount := 0, 0, 0
	for _, t := range p.tickets {
		switch t.DetailLevel() {
		case vo.TicketDetailFull:
			fullCount++
		case vo.TicketDetailStandard:
			stdCount++
		case vo.TicketDetailStub:
			stubCount++
		}
	}

	var b strings.Builder
	fmt.Fprintf(&b, "Ticket Plan: %s\n", p.planID)
	fmt.Fprintf(&b, "Epics: %d\n", len(p.epics))
	fmt.Fprintf(&b, "Tickets: %d (FULL=%d, STANDARD=%d, STUB=%d)\n\n",
		len(p.tickets), fullCount, stdCount, stubCount)

	for _, epic := range p.epics {
		fmt.Fprintf(&b, "  %s (%s):\n", epic.Title(), strings.ToUpper(string(epic.Classification())))
		for _, ticket := range p.tickets {
			if ticket.EpicID() == epic.EpicID() {
				fmt.Fprintf(&b, "    - [%s] %s\n", string(ticket.DetailLevel()), ticket.Title())
			}
		}
		b.WriteString("\n")
	}
	return b.String(), nil
}

// PromoteStub promotes a STUB ticket to FULL detail.
func (p *TicketPlan) PromoteStub(ticketID string, profile vo.StackProfile) error {
	for i, ticket := range p.tickets {
		if ticket.TicketID() == ticketID {
			if ticket.DetailLevel() != vo.TicketDetailStub {
				return fmt.Errorf("ticket '%s' is %s, not STUB -- cannot promote: %w",
					ticketID, string(ticket.DetailLevel()), domainerrors.ErrInvariantViolation)
			}
			resolvedProfile := profile
			if resolvedProfile == nil {
				resolvedProfile = p.profile
			}
			if resolvedProfile == nil {
				resolvedProfile = vo.PythonUvProfile{}
			}
			agg := vo.NewAggregateDesign(
				ticket.AggregateName(), ticket.BoundedContextName(),
				ticket.AggregateName(), nil, nil, nil, nil,
			)
			newDesc := RenderTicketDetail(agg, vo.TicketDetailFull, resolvedProfile)
			p.tickets[i] = NewGeneratedTicket(
				ticket.TicketID(), ticket.Title(), newDesc,
				vo.TicketDetailFull, ticket.EpicID(),
				ticket.BoundedContextName(), ticket.AggregateName(),
				ticket.Dependencies(), ticket.Depth(),
			)
			return nil
		}
	}
	return fmt.Errorf("ticket '%s' not found: %w", ticketID, domainerrors.ErrInvariantViolation)
}

// Approve approves the plan (all or a subset), emitting TicketPlanApproved.
func (p *TicketPlan) Approve(approvedIDs []string) error {
	if p.approved {
		return fmt.Errorf("plan already approved: %w", domainerrors.ErrInvariantViolation)
	}
	if len(p.tickets) == 0 {
		return fmt.Errorf("cannot approve plan with no tickets: %w", domainerrors.ErrInvariantViolation)
	}

	allIDs := map[string]bool{}
	for _, t := range p.tickets {
		allIDs[t.TicketID()] = true
	}

	var finalApproved, finalDismissed []string
	if approvedIDs == nil {
		for _, t := range p.tickets {
			finalApproved = append(finalApproved, t.TicketID())
		}
	} else {
		for _, id := range approvedIDs {
			if !allIDs[id] {
				return fmt.Errorf("unknown ticket ID: %s: %w", id, domainerrors.ErrInvariantViolation)
			}
		}
		approvedSet := map[string]bool{}
		for _, id := range approvedIDs {
			approvedSet[id] = true
		}
		finalApproved = approvedIDs
		for id := range allIDs {
			if !approvedSet[id] {
				finalDismissed = append(finalDismissed, id)
			}
		}
	}

	p.approved = true
	p.events = append(p.events, NewTicketPlanApproved(p.planID, finalApproved, finalDismissed))
	return nil
}

// PromotionEligibleIDs returns IDs of STUB tickets whose dependencies are all resolved.
func (p *TicketPlan) PromotionEligibleIDs(resolvedIDs map[string]bool) map[string]bool {
	eligible := map[string]bool{}
	for _, ticket := range p.tickets {
		if ticket.DetailLevel() != vo.TicketDetailStub {
			continue
		}
		deps := ticket.Dependencies()
		if len(deps) == 0 {
			continue
		}
		allResolved := true
		for _, depID := range deps {
			if !resolvedIDs[depID] {
				allResolved = false
				break
			}
		}
		if allResolved {
			eligible[ticket.TicketID()] = true
		}
	}
	return eligible
}

// -- Private helpers ----------------------------------------------------------

func (p *TicketPlan) computeTicketDepths() (map[string]int, error) {
	if p.dependencyOrder == nil {
		return nil, fmt.Errorf("dependency order must be computed before depths: %w",
			domainerrors.ErrInvariantViolation)
	}

	allIDs := map[string]bool{}
	for _, t := range p.tickets {
		allIDs[t.TicketID()] = true
	}
	depsMap := map[string][]string{}
	for _, t := range p.tickets {
		var inPlanDeps []string
		for _, d := range t.Dependencies() {
			if allIDs[d] {
				inPlanDeps = append(inPlanDeps, d)
			}
		}
		depsMap[t.TicketID()] = inPlanDeps
	}

	depths := map[string]int{}
	for _, tid := range p.dependencyOrder.OrderedIDs() {
		maxDepth := -1
		for _, d := range depsMap[tid] {
			if dd, ok := depths[d]; ok && dd > maxDepth {
				maxDepth = dd
			}
		}
		if maxDepth >= 0 {
			depths[tid] = maxDepth + 1
		} else {
			depths[tid] = 0
		}
	}
	return depths, nil
}

func (p *TicketPlan) reclassifyByDepth(classificationByCtx map[string]vo.SubdomainClassification, profile vo.StackProfile) {
	depths, err := p.computeTicketDepths()
	if err != nil {
		return
	}

	updated := make([]GeneratedTicket, 0, len(p.tickets))
	for _, ticket := range p.tickets {
		depth := depths[ticket.TicketID()]
		classification := classificationByCtx[ticket.BoundedContextName()]
		tier := vo.ClassifyTier(depth, classification)
		newLevel := vo.TierToDetailLevel(tier, classification)

		if newLevel != ticket.DetailLevel() {
			agg := vo.NewAggregateDesign(
				ticket.AggregateName(), ticket.BoundedContextName(),
				ticket.AggregateName(), nil, nil, nil, nil,
			)
			newDesc := RenderTicketDetail(agg, newLevel, profile)
			updated = append(updated, NewGeneratedTicket(
				ticket.TicketID(), ticket.Title(), newDesc,
				newLevel, ticket.EpicID(),
				ticket.BoundedContextName(), ticket.AggregateName(),
				ticket.Dependencies(), depth,
			))
		} else {
			updated = append(updated, NewGeneratedTicket(
				ticket.TicketID(), ticket.Title(), ticket.Description(),
				ticket.DetailLevel(), ticket.EpicID(),
				ticket.BoundedContextName(), ticket.AggregateName(),
				ticket.Dependencies(), depth,
			))
		}
	}
	p.tickets = updated
}

func (p *TicketPlan) assignCrossBCDependencies(upstreamCtxs map[string]map[string]bool) {
	ticketsByCtx := map[string][]string{}
	for _, t := range p.tickets {
		ticketsByCtx[t.BoundedContextName()] = append(ticketsByCtx[t.BoundedContextName()], t.TicketID())
	}

	updated := make([]GeneratedTicket, 0, len(p.tickets))
	for _, ticket := range p.tickets {
		ctx := ticket.BoundedContextName()
		upstreams := upstreamCtxs[ctx]
		depIDs := make([]string, len(ticket.Dependencies()))
		copy(depIDs, ticket.Dependencies())

		for upstream := range upstreams {
			depIDs = append(depIDs, ticketsByCtx[upstream]...)
		}

		if len(depIDs) != len(ticket.Dependencies()) {
			updated = append(updated, NewGeneratedTicket(
				ticket.TicketID(), ticket.Title(), ticket.Description(),
				ticket.DetailLevel(), ticket.EpicID(),
				ticket.BoundedContextName(), ticket.AggregateName(),
				depIDs, ticket.Depth(),
			))
		} else {
			updated = append(updated, ticket)
		}
	}
	p.tickets = updated
}

func (p *TicketPlan) computeDependencyOrder() (DependencyOrder, error) {
	allIDs := map[string]bool{}
	for _, t := range p.tickets {
		allIDs[t.TicketID()] = true
	}

	inDegree := map[string]int{}
	dependents := map[string][]string{}
	for id := range allIDs {
		inDegree[id] = 0
		dependents[id] = nil
	}

	for _, ticket := range p.tickets {
		for _, depID := range ticket.Dependencies() {
			if allIDs[depID] {
				inDegree[ticket.TicketID()]++
				dependents[depID] = append(dependents[depID], ticket.TicketID())
			}
		}
	}

	// Kahn's algorithm
	var queue []string
	for id, deg := range inDegree {
		if deg == 0 {
			queue = append(queue, id)
		}
	}

	var ordered []string
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		ordered = append(ordered, current)
		for _, dep := range dependents[current] {
			inDegree[dep]--
			if inDegree[dep] == 0 {
				queue = append(queue, dep)
			}
		}
	}

	if len(ordered) != len(allIDs) {
		return DependencyOrder{}, fmt.Errorf("circular dependency detected in ticket plan: %w",
			domainerrors.ErrInvariantViolation)
	}

	return NewDependencyOrder(ordered), nil
}
